### ISSUE TITLE
Boolean type class

### ISSUE THREAD

**Issue by snowleopard** [haskell-foundation/foundation#86]
I often want `not`, `&&` and `||` to work with non `Bool` datatypes. Judging by how many lifted variants of these operators can be found in various libraries, this is a pretty common use-case. I've come across `.&.`, `.&&.`, `&&^`, and there may be others.

Perhaps, it makes sense to introduce a new type class, e.g. `Boolean`, so people no longer need to come up with Boolean-like operator names?
```haskell
class Boolean a where
    true   :: a
    false  :: a
    not    :: a -> a
    (&&)   :: a -> a -> a
    (||)   :: a -> a -> a

instance Boolean a => Boolean (b -> a) where ...
instance (Boolean a, Monad m) => Boolean (m a) where ... -- This one is a bit tricky actually
```
I am pretty sure the design space is big and one can easily overdesign it, so it becomes hard to use. But adding such a type-class outside Prelude/Foundation is painful as it requires hiding `not` and friends.

**Comment by vincenthz** [comment]
This is an interesting idea, but also prone to massive abuse:

```haskell
instance Boolean Int where
    true = 1
    false = 0
    not = (/=) 0
    ...

when 0 $ do ...
when myPtr $ do ..
```

There's also `if` that work exclusively with `Bool`.

Maybe if we could just define an isomorphism abstraction, to make sure only isomorphic type (e.g. flag type, sum type with only 2 elements, ..)

Anyone got opinions about this ?

**Comment by vincenthz** [comment]
Also I have to say that in certain case `when myPtr ...` could be quite idiomatic (specially if you're rewriting lowlevel code) and useful, but also feel like a massive abuse of "everything-is-a-x" present in many weak type system.

**Comment by NicolasDP** [comment]
Indeed, the System programmer in me would like this as well.

I think `when` and `unless` would make a better use of:

```haskell
class Boolean a where isTrue a :: Bool
when :: (Applicative f, Boolean a) => a -> f () -> f ()
unless b f = when (not $ isTrue b) f
```

It seems to me, that @snowleopard is proposing to extend `Alternative` with a `not` operation:

```haskell
class Alternative f => Boolean f where
    not    :: a -> a

true :: Boolean f => f
true   = pure
false :: Boolean f => f
false  = empty
(&&) :: Boolean f => f -> f -> f
(&&)  = (*>)
(||) :: Boolean f => f -> f -> f
(||)   = (<|>)
```

**Comment by snowleopard** [comment]
@vincenthz I agree, this can be abused. Note though that `when 0` is already possible:
```haskell
instance Num Bool where
    fromInteger = (/= 0)

when 0 $ do ...
```

To minimise abuse we could keep `when` non-generalised: `when :: Bool -> m a -> m a`. I think the main benefit from the Boolean type class is not at use sites like `if` or `when`, but in building Boolean expressions with unified operators.

> Maybe if we could just define an isomorphism abstraction, to make sure only isomorphic type (e.g. flag type, sum type with only 2 elements, ..) are used through this facility

Unfortunately this disallows many useful instances, e.g. those that I listed: `Boolean a => Boolean (b -> a)`, `(Boolean a, Monad m) => Boolean (m a)`. 

By the way, two other potential problems with this proposal are: 1) as any generalisation this makes the language less friendly for newcomers, 2) Error messages may become more complicated.

**Comment by snowleopard** [comment]
> It seems to me, that @snowleopard is proposing to extend `Alternative` with a `not` operation:

@NicolasDP No, this wasn't on my mind :-) Not sure I fully understand the interplay between `Alternative` and `Boolean`, but this seems to be a plausible instance too.

**Comment by vincenthz** [comment]
Ok, I think I'm fine if we restrict this to building relation. Need to see the full extend of this exercise, but I guess we can see if that's going to fly now.

Related to errors and newcomers, I think if we documents things properly somewhere, then it should be ok. I think a set of recipes for common problems that can be found in foundation generalization of things would be awesome. Also from GHC 8.0 we might have some help from the error extension.

Not very keen on `Alternative` though, as the concept are different (I appreciate there's some similarity)

**Comment by snowleopard** [comment]
@vincenthz Shall I prepare a draft PR or would you like to take the lead on this?

**Comment by vincenthz** [comment]
I'm off for a week, so I'm not going to do anything right now (and be quite irresponsive for a bit). if you would prepare a draft PR, it's probably best. If anyone want to further comments on this also, more than welcome.

**Comment by snowleopard** [comment]
We need to distinguish two different concepts here: _Boolean algebra_ and _Boolean values_ (i.e. bits 0 or 1). The latter form a Boolean algebra, but in general a Boolean algebra can have more than 2 values, e.g. integers form a Boolean algebra under bitwise operations.

To be clear: **initially this proposal was for a type class for Boolean algebra**. Note: `if` and `when` only make sense for Boolean values and are hence not part of this proposal.

Looking at @ndmitchell's `Control.Monad.Extra` [library](https://hackage.haskell.org/package/extra-1.5/docs/Control-Monad-Extra.html) I see that `(&&^)` and friends operate on Boolean values in a monad (the implementation relies on the fact that Boolean values can be tested, i.e. converted to `Bool`), therefore this proposal will not cover it either.

So, maybe we should be talking about two type classes: `Boolean` and `Bit`? The former provides a unified interface for Boolean operations, while the latter provides a unified interface for Boolean values, which are commonly known as _bits_ (this is what @vincenthz was alluding to when mentioning flags and in general things isomorphic to `Bool`)?

In this setting `Bool` will be an instance of both, so it can be manipulated by `(&&)` etc. and also tested, for example, via a standard function `bool :: a -> a -> Bool -> a` that could be a member of the type class `Bit`. In principle one could also allow any `Bit` values to be used in `when` etc.

**Comment by NicolasDP** [comment]
Ok, I see now. Please, feel free to go ahead and prepare a draft PR.
